File: Yandex/Practicum/Algorightms_and_Data_Structures/3_sprint/N.Клумбы/solution.cpp
#include <iostream>
#include <vector>
#include <algorithm>


void solution(int& n, std::vector<std::pair<int, int>>& pairs) {
    /*
    Time Complexity O(NlogN)
    Memory Complexity O(1)
    */
    std::sort(pairs.begin(), pairs.end());
    int mainIdx = 0;
    int currentIdx = 0;
    for (int idx = 0; idx < n; ++idx) {
        if (pairs[currentIdx].second >= pairs[idx].first) {
            pairs[currentIdx].second = std::max(pairs[currentIdx].second, pairs[idx].second);
        } else {
            pairs[mainIdx] = pairs[currentIdx];
            currentIdx = idx;
            ++mainIdx;
        }
    }
    pairs[mainIdx] = pairs[currentIdx];
    ++mainIdx;
    for (int idx = 0; idx < mainIdx; ++idx) {
        std::cout << pairs[idx].first << " " << pairs[idx].second << "\n";
    }
}


int main() {
    int n;
    std::cin >> n;
    std::vector<std::pair<int, int>> pairs;
    for (int i = 0; i < n; ++i) {
        int start, end;
        std::cin >> start >> end;
        pairs.push_back({start, end});
    }
    solution(n, pairs);
}

File: Yandex/Practicum/Algorightms_and_Data_Structures/3_sprint/N.Клумбы/solution.py
import sys


def solution(n: int, pairs: list[list[int, int]]) -> None:
    """
    Time Complexity O(NlogN)
    Memory Complexity O(1)
    """
    pairs.sort()
    main_idx: int = 0
    current_idx: int = 0
    for idx in range(n):
        if pairs[current_idx][1] >= pairs[idx][0]:
            pairs[current_idx][1] = max(pairs[current_idx][1], pairs[idx][1])
        else:
            pairs[main_idx] = pairs[current_idx]
            current_idx = idx
            main_idx += 1
    pairs[main_idx] = pairs[current_idx]
    main_idx += 1
    for idx in range(main_idx):
        sys.stdout.write(' '.join(map(str, pairs[idx])) + '\n')


if __name__ == '__main__':
    n: int = int(sys.stdin.readline().rstrip())
    pairs: list[list[int, int]] = []
    for _ in range(n):
        start, end = map(int, sys.stdin.readline().rstrip().split())
        pairs.append([start, end])
    solution(n, pairs)